### ISSUE TITLE
train and my dataset

### ISSUE THREAD

**Issue by lzbushicai** [cv-rits/MonoScene#85]
Thank you for your outstanding work. I want to use your model to train my dataset. My dataset contains RGB images and semantic information. how can  process my dataset? Also, how can I use your existing model to detect  my images?

my dataset tree is |-- datasets
 |  |-- ORFD
 |  |  |-- training
 |  |  |  |-- sequence   |-- calib
 |  |  |                 |-- sparse_depth
 |  |  |                 |-- dense_depth
 |  |  |                 |-- lidar_data
 |  |  |                 |-- image_data
 |  |  |                 |-- gt_image
 ......
 |  |  |-- validation
 ......
 |  |  |-- testing
 ......

I don"t have voxeldata can I train with your monoscene?

**Comment by anhquancao** [comment]
Hi @lzbushicai, you can voxelize your lidar into voxel grid and adapt the [kitti dataloader](https://github.com/astra-vision/MonoScene/blob/master/monoscene/data/semantic_kitti/kitti_dataset.py#L68) to your data.

**Comment by lzbushicai** [comment]
Thank you for the reply! ! ! Now I want to perform semantic completion work in the wild scene (ORDF dataset), but currently I only have RGB image and point cloud data. KITTI's VOXEL folder has four suffix files: **.bin .valied .occloud .label** They seem to have marked the point cloud files. I only have .bin lidar data now. I don’t know if there is any difference between the .bin file under my lidar_data and the _/KITTI/data_odometry_voxels/dataset/sequences/00/0000.bin_ file, as well as my How lidar_data is voxelized into the organizational form of data_odometry_voxels. There is another question. I also downloaded the NYU data set, but there are only .bin files, RGB image and depth information in NYU, and no voxeldata. but your nyu/process.py convert  .bin file to voxel,it seems that the NYU offical dataset don’t have .bin file，  How did you  generate the .bin file and get the voxel labels?
Is it feasible for my data set to work with monoscene?
  
I found the inference code https://huggingface.co/spaces/CVPR/MonoScene/blob/main/app.py#L29 in the previous issue. I run app.py and go to Running on local URL:  http://127.0.0.1:7860
,but there is no responce to my inference 
![image](https://github.com/astra-vision/MonoScene/assets/114228032/0f3e3a18-224e-4f50-9bc0-95497d38a8d1)
what I can do to solve this problem？

I am looking forward your reply!!!

**Comment by anhquancao** [comment]
Hi, I use this code to convert lidar to voxelized scene
```
voxel_size = 0.2
vox_origin = np.array([0, -25.6, -2]).reshape(1, 3)
vox_origin = self.vox_origin.reshape(1, 3)
voxels = (xyz - vox_origin.reshape(1, 3)) // self.voxel_size
```
But if you only need inference, you don't need the point cloud. You can take a look directly at the inference code of the Gradio demo. However, I don't know why the UI doesn't work any more. 

For NYUv2, I took the repacked dataset from the previous work
https://github.com/charlesCXK/TorchSSC

**Comment by lzbushicai** [comment]
thank you！！！I am trying to  make my liar label,I need train my dataset,thank you for your reply!!

**Comment by anhquancao** [comment]
You're welcome! Good luck with your work. Feel free to reopen if you encounter another issue.

**Comment by lzbushicai** [comment]
It seems there is a small problem in https://huggingface.co/spaces/CVPR/MonoScene/blob/27c712a835068b81df1b44c9c928af9fd1c56674/helpers.py#L42and L86 show show opposite coordinate transformations and in L144 function get_projections cam_E ，The variableT_velo_2_cam is a transformation from radar system to camera system, but the annotation is a transformation from camera system to radar system，which is conflict

**Comment by anhquancao** [comment]
Sorry, I don't understand. Function vox2pix generates voxel grid in lidar coordinates and project them to the image. What do you mean by opposite coordinate transformations?

**Comment by lzbushicai** [comment]
sorry,I forget to reply, I mean I don't understand the projection of lidar coordinates from image coordinats and voxel cooredinats,I don't understand the projection process in the middle

**Comment by anhquancao** [comment]
Sorry, I don't know what to answer, could you be more specific?

**Comment by lzbushicai** [comment]
> Sorry, I don't know what to answer, could you be more specific?

What I mean is, in the **vox2pix** function of **helpers. py,** according to the meaning of your code, **cam_ pts=fusion. rigid_Transform (cam_pts, cam_E)** is the translation of voxels from the lidar coordinate system to the camera coordinate system, but your annotation under **vox2pix** states "cam-E: **4x4"=Camera pose in case of NYUv2 dataset=Transformation from camera to lidar coordinate in case of SemKITTI**, explain cam_E is from the camera system to the lidar system. May I ask if I misunderstood or if your expression was incorrect?

**Comment by lzbushicai** [comment]
Hi,Cao Anh Quan
Hello! My **velodyne and labels** data have point cloud and point cloud label files in the same format as SemanticKitti in. bin and. label formats. The following image shows the result of visualizing SemanticKitti and my dataset bin files.



sequence003_ 116

I organized my files into the format shown below and successfully generated a voxelated. bin. label. occloud. invalid file using https://github.com/jbehley/voxelizer





But I combined my voxel file and SemanticKITTI's voxel file, using https://github.com/PRBonn/semantic-kitti-api/blob/master/visualize_voxels.py Visualize it, but the effect is very strange. The left side of the image shows my data. On the right is Kitti's data



I am confident that my original bin and labels are fine. Is there a problem with my configuration? The following image shows my configuration



In addition, I also found that the sequence numbers of SemanticKITTI's voxel data are multiples of 5, indicating that SemanticKITTI ignored the first 4 frames. Can you teach me how to set them? I want to change my voxel to SemanticKITTI format. How can I modify my configuration? Can we generate voxel files with five intervals below?

In other words, how can I verify the correctness of the label file in the voxel folder? or How can I visualize it?  I read the code for voxel/label as follows
`rellis_label = "/home/liuzhen/learning/DATASETS/RELLIS3D/voxelizer/sequences/00/voxels/000003.label"
    voxels = np.fromfile(rellis_label, dtype=np.uint16).reshape((256, 256, 32))`

**Comment by lzbushicai** [comment]
Hi,Cao Anh Quan
My** velodyne **and** labels** data have point cloud and point cloud label files in the same format as SemanticKitti **in. bin and. label formats**. The following image shows the result of visualizing SemanticKitti and my dataset bin files.

![](https://secure2.wostatic.cn/static/gXJBynRhjWbVM7AkcKUrT6/image.png?auth_key=1705749283-wZE9B9UNCZvNJ45fBRKgGc-0-247b170e783dcdbbc9a3d62915944a2e)



![sequence003_ 116](https://secure2.wostatic.cn/static/sF6FJTSHGgTSe45bx1U3fb/image.png?auth_key=1705749290-9oWqFtq1AV1mWfJ6K6FaDx-0-fb9b1b89ed03536faa6dfa7491d4b77a)

I organized my files into the format shown below and successfully generated a voxelated. bin. label. occloud. invalid file using (https://github.com/jbehley/voxelizer)

![](https://secure2.wostatic.cn/static/xd1fbpRqjHwBpj7xaM8DBW/image.png?auth_key=1705749350-qFqHuCXCJxreBmj1BSohhM-0-a6ef44d376c259d026dbc68c4d214652)

![](https://secure2.wostatic.cn/static/t7BgmanjZcchciEV2HMvio/image.png?auth_key=1705749341-dpy88MZam3PWpHGUUUyhtw-0-00f1f7d2b03cdfb5d3446a21e7199852)

But I combined my voxel file and SemanticKITTI's voxel file, using [https://github.com/PRBonn/semantic-kitti-api/blob/master/visualize_voxels.py](https://github.com/PRBonn/semantic-kitti-api/blob/master/visualize_voxels.py) Visualize it, but the effect is very strange. The left side of the image shows my data. On the right is Kitti's data

![](https://secure2.wostatic.cn/static/kjFZiPR3UuT1sdeuAfqXUZ/Kitti_rellis_label.png?auth_key=1705749403-8tRUzhHQtxKmqrkH1gJsNt-0-ef74e08134506387458bfd017543ab11)

I am confident that my original bin and labels are fine. Is there a problem with my configuration? The following image shows my configuration

![](https://secure2.wostatic.cn/static/fhuUeGWc2jAwwKcPodQnmi/image.png?auth_key=1705749432-5ejEALseEnF2uh5XAmkdKq-0-17fce5b0184d569f41c8e485e757420a)

In addition, I also found that the sequence numbers of SemanticKITTI's voxel data are multiples of 5, indicating that SemanticKITTI ignored the first 4 frames. Can you teach me how to set them? I want to change my voxel to SemanticKITTI format. How can I modify my configuration? Can we generate voxel files with five intervals below?

![](https://secure2.wostatic.cn/static/kHPiAzSkxF2eWnmzGGixhf/image.png?auth_key=1705749455-fNNtq1HPzbRRjqvGgQQVWo-0-1366903dc8345ff5b958816e97151b52)
In addition, I also found that the sequence numbers of SemanticKITTI's voxel data are multiples of 5, indicating that SemanticKITTI ignored the first 4 frames. Can you teach me how to set them? I want to change my voxel to SemanticKITTI format. How can I modify my configuration? Can we generate voxel files with five intervals below?
` my_label = "/home/liuzhen/learning/DATASETS/voxelizer/sequences/00/voxels/000003.label"
    voxels = np.fromfile(rellis_label, dtype=np.uint16).reshape((256, 256, 32))`

**Comment by lzbushicai** [comment]
> Hi,Cao Anh Quan My** velodyne **and** labels** data have point cloud and point cloud label files in the same format as SemanticKitti **in. bin and. label formats**. The following image shows the result of visualizing SemanticKitti and my dataset bin files.
> 
> ![](https://camo.githubusercontent.com/49a6bfb24e7a49bd9938d077f88bc763f6d8a427b1b350af672b299d0c61e7ba/68747470733a2f2f736563757265322e776f7374617469632e636e2f7374617469632f67584a42796e52686a5762564d37416b634b557254362f696d6167652e706e673f617574685f6b65793d313730353734393238332d775a45394239554e435a764e4a34356642524b6747632d302d3234376231373065373833646364626263396133643632393135393434613265)
> 
> ![sequence003_ 116](https://camo.githubusercontent.com/f928097afffd4010f74ad9973ef7c3cc66457543ba498e3a733816b2db2a882c/68747470733a2f2f736563757265322e776f7374617469632e636e2f7374617469632f734636464a54534847675453653435627831553366622f696d6167652e706e673f617574685f6b65793d313730353734393239302d396f5771467471314156316d57664a364b36466144782d302d6662396231623839656430333533366661613664666137343931643462373761)
> 
> I organized my files into the format shown below and successfully generated a voxelated. bin. label. occloud. invalid file using (https://github.com/jbehley/voxelizer)
> 
> ![](https://camo.githubusercontent.com/1f2b671d0220f6a4b5e1f045cff8b120ee0cb6428c916a1658570d1bde9a86f8/68747470733a2f2f736563757265322e776f7374617469632e636e2f7374617469632f78643166627052716a487742706a3778614d384442572f696d6167652e706e673f617574685f6b65793d313730353734393335302d71467148754358434a787265426d6a3142536f68684d2d302d6136656634346433373663323539643032366462633638633464323134363532)
> 
> ![](https://camo.githubusercontent.com/bb7da834e7aa49ec6f0eff98f6b7f45d9d581371ae796316147dd3c75ae7eed2/68747470733a2f2f736563757265322e776f7374617469632e636e2f7374617469632f743742676d616e6a5a6363686369455632484d76696f2f696d6167652e706e673f617574685f6b65793d313730353734393334312d64707938384d5a616d335057704847555555796874772d302d3030663166376432623033636466623564333434366132316537313939383532)
> 
> But I combined my voxel file and SemanticKITTI's voxel file, using https://github.com/PRBonn/semantic-kitti-api/blob/master/visualize_voxels.py Visualize it, but the effect is very strange. The left side of the image shows my data. On the right is Kitti's data
> 
> ![](https://camo.githubusercontent.com/fa7576e959b8e080c85324b6ff9a7f2f7bf6ef312fcb6882734bab4a9c765893/68747470733a2f2f736563757265322e776f7374617469632e636e2f7374617469632f6b6a465a69505233557554317364657541667158555a2f4b697474695f72656c6c69735f6c6162656c2e706e673f617574685f6b65793d313730353734393430332d387452557a68485174784b6d71726b4831674a734e742d302d6566373465303831333435303633383734353862666430313735343361623131)
> 
> I am confident that my original bin and labels are fine. Is there a problem with my configuration? The following image shows my configuration
> 
> ![](https://camo.githubusercontent.com/61d35042c0ed11cba0f192ba5d6438cafe6a3cc91e91be02e07b9a32e51e9931/68747470733a2f2f736563757265322e776f7374617469632e636e2f7374617469632f6668755565475763326a4177774b63506f64516e6d692f696d6167652e706e673f617574685f6b65793d313730353734393433322d35656a45414c7365456e463275683558416d6b644b712d302d3137666365356230313834643536396634316338653438356537353734323061)
> 
> In addition, I also found that the sequence numbers of SemanticKITTI's voxel data are multiples of 5, indicating that SemanticKITTI ignored the first 4 frames. Can you teach me how to set them? I want to change my voxel to SemanticKITTI format. How can I modify my configuration? Can we generate voxel files with five intervals below?
> 
> ![](https://camo.githubusercontent.com/47dad65b81b8baa83834671c65ca74e09228a0f65d58a3672ae6f0c5cc2277b4/68747470733a2f2f736563757265322e776f7374617469632e636e2f7374617469632f6b485069417a536b78463265576e6d7a4747697868662f696d6167652e706e673f617574685f6b65793d313730353734393435352d664e4e74713148507a6252526a71764767515156576f2d302d3133363639303364633833343566663562393538383136653937313531623532) In addition, I also found that the sequence numbers of SemanticKITTI's voxel data are multiples of 5, indicating that SemanticKITTI ignored the first 4 frames. Can you teach me how to set them? I want to change my voxel to SemanticKITTI format. How can I modify my configuration? Can we generate voxel files with five intervals below? ` my_label = "/home/liuzhen/learning/DATASETS/voxelizer/sequences/00/voxels/000003.label" voxels = np.fromfile(rellis_label, dtype=np.uint16).reshape((256, 256, 32))`

Hi,Cao Anh Quan！I have received a response from the original author, and my problem has been resolved. Thank you!, Here is the author's response:
**Here a possible things that you should check, such that we can rule out "easy problems":

Check if you have maybe a wrong file (you cannot just copy the KITTI file as we opted to use the KITTI way of representing poses (i.e., all poses are given in camera coordinates). If you have simply the with LiDAR poses, then you should use a with identity matrices.calib.txtposes.txtcalib.txt
It would help to have not only a comparison what comes out with SemanticKITTI (i.e., having the original scan from KITTI; at least the point cloud looks like sequence 00 from KITTI), but also your data. At least that would provide an indication of what your are expecting to see.
Regarding your other questions:

You can generate also "dense" voxels, i.e., voxel grids for each and every scan. But due to space limitations, we decided to use only every 5th to make the download and the prediction more manageable. In the voxelizer there should be also an option to have voxel grids for each an every scan.
How many scans are used to aggregate for the scan is given by prior and past scans (that means how many scans should be accumulated); Even if we generate the voxel grids only for every 5th timestamp, we still aggregate all prior scans (I think it was something like 100 scans? for the ground truth). In your shown configuration, you now take "only" the last 4 before the current timestamp and the future 4 scans. I agree, the naming is a bit suboptimal (prior = before the current timestamp, past = after the current timestamp).**

**Comment by anhquancao** [comment]
Dear @lzbushicai, I am pleased to hear that. Best wishes for your research.

**Comment by lzbushicai** [comment]
Dear Cao Anh Quan！
I meet I encountered other difficulties during my training。

I set **relation_loss:**  true，**sem_scal_loss**: true，**geo_scal_loss**: true,the program will execute normally，but when I set **CE_ssc_loss** true。The program will report the following error：
`Validation sanity check:  50%|██████████          | 1/2 [00:20<00:20, 20.78s/it]../aten/src/ATen/native/cuda/NLLLoss2d.cu:104: nll_loss2d_forward_kernel: block: [9,0,0], thread: [897,0,0] Assertion t >= 0 && t < n_classes failed.
../aten/src/ATen/native/cuda/NLLLoss2d.cu:104: nll_loss2d_forward_kernel: block: [9,0,0], thread: [899,0,0] Assertion t >= 0 && t < n_classes failed.
../aten/src/ATen/native/cuda/NLLLoss2d.cu:104: nll_loss2d_forward_kernel: block: [9,0,0], thread: [259,0,0] Assertion t >= 0 && t < n_classes failed.
../aten/src/ATen/native/cuda/NLLLoss2d.cu:104: nll_loss2d_forward_kernel: block: [9,0,0], thread: [260,0,0] Assertion t >= 0 && t < n_classes failed.
../aten/src/ATen/native/cuda/NLLLoss2d.cu:104: nll_loss2d_forward_kernel: block: [9,0,0], thread: [929,0,0] Assertion t >= 0 && t < n_classes failed.
../aten/src/ATen/native/cuda/NLLLoss2d.cu:104: nll_loss2d_forward_kernel: block: [9,0,0], t`
I guess the tag contains invalid values,is me right?In order to adapt to my data, I have changed `monoscene/data/semantic_kitti/semantic-kitti.yaml` to the following format
```
`nbr_classes: 20
grid_dims: [256, 32, 256]  # (W, H, D)
labels:
  0: "void"
  1: "dirt"
  3: "grass"
  4: "tree"
  5: "pole"
  6: "water"
  7: "sky"
  8: "vehicle"
  9: "object"
  10: "asphalt"
  12: "building"
  15: "log"
  17: "person"
  18: "fence"
  19: "bush"
  23: "concrete"
  27: "barrier"
  31: "puddle"
  33: "mud"
  34: "rubble"
color_map: # bgr
  0: [ 0, 0, 0 ]
  1: [ 108, 64, 20 ]
  3: [ 0, 102, 0 ]
  4: [ 0, 255, 0 ]
  5: [ 0, 153, 153 ]
  6: [ 0, 128, 255 ]
  7: [ 0, 0, 255 ]
  8: [ 255, 255, 0 ]
  9: [ 255, 0, 127 ]
  10: [ 64, 64, 64 ]
  12: [ 255, 0, 0 ]
  15: [ 102, 0, 0 ]
  17: [ 204, 153, 255 ]
  18: [ 102, 0, 204 ]
  19: [ 255, 153, 204 ]
  23: [ 170, 170, 170 ]
  27: [ 41, 121, 255 ]
  31: [ 134, 255, 239 ]
  33: [ 99, 66, 34 ]
  34: [ 110, 22, 138 ]

content: # as a ratio with the total number of points
  0: 0.018889854628292943
  1: 0.0002937197336781505
  10: 0.040818519255974316
  11: 0.00016609538710764618
  13: 2.7879693665067774e-05
  15: 0.00039838616015114444
  16: 0.0
  18: 0.0020633612104619787
  20: 0.0016218197275284021
  30: 0.00017698551338515307
  31: 1.1065903904919655e-08
  32: 5.532951952459828e-09
  40: 0.1987493871255525
  44: 0.014717169549888214
  48: 0.14392298360372
  49: 0.0039048553037472045
  50: 0.1326861944777486
  51: 0.0723592229456223
  52: 0.002395131480328884
  60: 4.7084144280367186e-05
  70: 0.26681502148037506
  71: 0.006035012012626033
  72: 0.07814222006271769
  80: 0.002855498193863172
  81: 0.0006155958086189918
  99: 0.009923127583046915
  252: 0.001789309418528068
  253: 0.00012709999297008662
  254: 0.00016059776092534436
  255: 3.745553104802113e-05
  256: 0.0
  257: 0.00011351574470342043
  258: 0.00010157861367183268
  259: 4.3840131989471124e-05
# classes that are indistinguishable from single scan or inconsistent in
# ground truth are mapped to their closest equivalent
learning_map:
  0: 0
  1: 1
  3: 3
  4: 4
  5: 5
  6: 6
  7: 7
  8: 8
  9: 9
  10: 10
  12: 11
  15: 15
  17: 17
  18: 18
  19: 19
  23: 23
  27: 27
  31: 31
  33: 33
  34: 34  # "moving-other"-vehicle to "other-vehicle" ----------------mapped
learning_map_inv: # inverse of previous map
  0: 0
  1: 1
  3: 3
  4: 4
  5: 5
  6: 6
  7: 7
  8: 8
  9: 9
  10: 10
  12: 11
  15: 15
  17: 17
  18: 18
  19: 19
  23: 23
  27: 27
  31: 31
  33: 33
  34: 34     # "traffic-sign"
learning_ignore: # Ignore classes
  0: True      # "unlabeled", and others ignored
  1: False     # "car"
  3: False     # "bicycle"
  4: False     # "motorcycle"
  5: False     # "truck"     # "other-vehicle"
  6: False     # "person"
  7: False     # "bicyclist"
  8: False     # "motorcyclist"
  9: False     # "road"
  10: False    # "parking"
  12: False    # "sidewalk"
  15: False    # "other-ground"
  17: False    # "building"
  18: False    # "fence"
  19: False    # "vegetation"
  23: False    # "trunk"
  27: False    # "terrain"
  31: False    # "pole"
  33: False
  34: False     # "traffic-sign"
split: # sequence numbers
  train:
    - 0
    - 1
    - 2
    - 3
  valid:
    - 4
  test:
    - 4
```
Where did I go wrong?

another,Regarding the production of the demo video, I think I can visualize my voxels using **monoscene/scripts/visualization/kitti_vis_pred.py.**py and then concatenate them frame by frame. Do you think what I said is correct?

**Comment by lzbushicai** [comment]
awsdg

**Comment by anhquancao** [comment]
- You should check if the length class_weights is equal to your number of classes.
https://github.com/astra-vision/MonoScene/blob/master/monoscene/loss/ssc_loss.py#L95

- Regarding visualization, I draw and visualize them frame by frame.

**Comment by lzbushicai** [comment]
Thank you for your reply. I am very sorry for replying so late because I am on winter vacation. I will check the suggestions you gave
